File: 스택의 활용/JunYoungKr/9012.py
from collections import deque
import sys

input = sys.stdin.readline

# N 입력
N = int(input())

for _ in range(N):
    arr = list(input().strip())
    
    stack = []
    flag = True
    
    # 여는 괄호라면 받음
    for e in arr:
        if e in '([':
            stack.append(e)
            continue
        
        # stack의 길이가 0이 아닌 경우
        if e in ')]':
            if len(stack) == 0:
                flag = False
                break
            # 마지막 요소가 여는 괄호인데
            elif len(stack) != 0:
                마지막요소 = stack.pop()
                if 마지막요소 == '(' and e == ']':
                    flag = False
                    break
                elif 마지막요소 == '[' and e == ')':
                    flag = False
                    break
                
                
     # stack의 길이가 0인 경우
    if len(stack) >= 1:
        flag = False
    
    if flag:
        print("YES")
    else:
        print("NO")
            
            

File: 스택/rian00/9012.js
const { sep } = require("path");

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

const [N, ...arr] = input;

for (const str of arr) {
  const stack = [];
  let flag = false;
  for (const e of str) {
    flag = false;
    switch (e) {
      case "(":
      case "[":
        stack.push(e);
        break;
      case ")":
        if (stack.length > 0 && stack[stack.length - 1] === "(") {
          stack.pop();
        } else {
          console.log("NO");
          flag = true;
        }
        break;
      case "]":
        if (stack.length > 0 && stack[stack.length - 1] === "[") {
          stack.pop();
        } else {
          console.log("NO");
          flag = true;
        }
        break;
    }
    if (flag) break;
  }
  if (!flag && stack.length === 0) console.log("YES");
  else if (!flag && stack.length > 0) console.log("NO");
}
